Firestore/core/src/firebase/firestore/remote/stream.mm

```diff
@@ -228,41 +223,64 @@
 
 void Stream::Stop() {
   EnsureOnQueue();
-
   LOG_DEBUG("%s stop", GetDebugDescription());
 
-  if (!IsStarted()) {
+  Close(Status::OK());
+}
+
+void Stream::Close(const Status& status) {
+  // This function ensures that both graceful stop and stop due to error go
+  // through the same sequence of steps. While it leads to more conditional
+  // logic, the benefit is reducing the chance of divergence across the two
+  // cases.
+
+  EnsureOnQueue();
+  bool graceful_stop = status.ok();
+
+  // Step 1 (both): check current state.
+  if (graceful_stop && !IsStarted()) {
+    // Graceful stop is idempotent.
     return;
   }
+  HARD_ASSERT(IsStarted(), "Trying to close a non-started stream");
 
-  // If the stream is in the auth stage, gRPC stream might not have been created
-  // yet.
-  if (grpc_stream_) {
+  // Step 2 (both): cancel any outstanding timers (they're guaranteed not to
+  // execute).
+  CancelIdleCheck();
+  backoff_.Cancel();
+
+  // Step 3 (both): increment close count, which invalidates the auth callback,
+  // guaranteeing it won't execute.
+  ++close_count_;
+
+  // Step 4 (both): make small adjustments (to backoff/etc.) based on the
+  // status.
+  if (graceful_stop) {
+    // If this is an intentional close, ensure we don't delay our next
+    // connection attempt.
+    backoff_.Reset();
+  } else {
+    HandleErrorStatus(status);
+  }
+
+  // Step 5 (graceful stop only): gracefully finish the underlying stream.
```

Comment:
nit: there's a lot of "graceful" in this comment below. Maybe: "Step 5 (graceful stop only): give subclasses a chance to send final messages.

Reply:
Done.